### ISSUE TITLE
Specifying Isomer Type for Glycan Residues?

### ISSUE THREAD

**Issue by Monikerton** [ibmm-unibe-ch/glycosylator#2]
Description:
While trying to create a specific glycan chain using IUPAC notation with Glycosylator, the module returns incorrect isomers for different glycans.

Example:
```
import glycosylator as gl
iupac = "Glc(b1-3)Fuc"
glycan = gl.glycan(iupac)
print(glycan)
--------------------------
Molecule None:
╙── Residue(FCA, 1)
    └── Residue(BGC, 2)
```

it returns [FCA](https://www.rcsb.org/ligand/FCA) in place of [FUC](https://www.rcsb.org/ligand/FUC), which are isomers of each other, and [BGC](https://www.rcsb.org/ligand/BGC) in place of [GLC](https://www.rcsb.org/ligand/GLC), which are also isomers. 

Analysis:
Upon reviewing the source code of _parse_iupac_graph in Glycan.py, it appears that if multiple results are found for the same molecule, the function returns the first result found (lines 544 and 560 when I view it on vscode).

Request:
Is there a way to specify the exact isomer to return when multiple results are available?

**Comment by NoahHenrikKleinschmidt** [comment]
Hi @Monikerton 
I see. Yes we can do something about that, I think. 
In general, glycosylator tries to find the right isomer based on the specification of the linkage - e.g. `BGC` to have the O1 in the equatorial position for a beta linkage or `GLC` for the axial position of an alpha linkage. So, when specifying `Glc(b1-3)Fuc` glycosylator knows it has to find a glucose and which one will be determined by  the `(b` or `(a` that precedes it. That's why you are getting `BGC` instead of `GLC` in your case. As for the `FCA` instead of `FUC`, you are right, that comes simply from the fact that multiple isomers are in the database and the first one is returned. If you want to control which isomer is used you can simply pass the PDB-id instead of the IUPAC abbreviation directly like so:

```python
>>> iupac = "Glc(b1-3)FUC`
>>> glycan = gl.glycan(iupac)
>>> print(glycan)
Molecule None:
╙── Residue(FUC, 1) # <- now a FUC instead of FCA
    └── Residue(BGC, 2)
```

Note however, that when I've tested this for the glucose it still took the BGC since the primary method for identifying isomers is from the linkage specification. However, I am thinking we could change that to check whether a PDB-ID in all uppercase is provided and in this case to directly take that compound instead of relying on the linkage to discern which isomer to use. 

What do you think about this?

**Comment by Monikerton** [comment]
I think that’s an excellent idea. In that case, I would also include a warning/check that specifies whether it would be stereo chemically valid for that molecule to create the bond. As you mentioned, I thought GLC was the appropriate isomer to use in this structure, but you are right that it should be BGC. For reference: https://glycoshape.io/glycan?IUPAC=Glc(b1-3)Fuc . I would also note that when searching for BGC, the code also returned Z8T as a potential candidate. 

Regardless, I think the option to force glycosylator to include a specific molecule would be good.

**Comment by NoahHenrikKleinschmidt** [comment]
Great then look forward to the next release (or sooner on the `dev` branch) 👍 - I'll see what we might do about warnings, I'm not a fan of them though. Code should either crash directly or run without a fuzz in my opinion 😅  

Regarding the BGC and Z8T, the difference is that BGC is the $D$ configuration and Z8T is $L$. In the database, both answer to the IUPAC identifier `b-Glc`, so they are both returned. If you want to make sure you have one isomer specifically you can add this into the IUPAC string by adding `l-` or `d-` in front of the monosaccharide code. This usually requires that you also add `p` or `f` (for pyranose or furanose) after the monosaccharide code like so:

```python
# standard
iupac_default = "Gal(b1-4)GlcNAc"

# specifying to use the Beta-L Galactopyranose
iupac_specific = "l-Galp(b1-4)GlcNAc"

# literally giving the PDB id of the component to use
iupac_with_pdbIDs = "GIV(b1-4)GlcNac"
```

PS: That's not a naming rule that I made up; that's literally how the data attributes are handled by the PDB small component's dictionary, which is where Glycosylator gets the structures and metadata from. 

I think I'm going to add stuff about this to the tutorials so that people know about it 🤔

**Comment by NoahHenrikKleinschmidt** [comment]
Alright,
So I made the changes and it should be fine in the new release `v.5.6.3` now. Also, you can check out the [IUPAC to Glycan tutorial](https://glycosylator.readthedocs.io/en/latest/examples/building_glycan_from_iupac.html) where I've added a new section to describe how to specify isomers and even non-canonical pseudo-glycans. 

Hopefully this fixes your issue. Feel free to reopen the issue if there is still some stuff not working 😄

**Comment by Monikerton** [comment]
Thanks so much for all the work you’ve done, the developments you've made, and being so prompt in responding. I have another question concerning specifying glycans:

I tried using the format you mentioned, and it worked fine for the three-letter codes. However, I encountered an issue when specifying isomers for N-Acetylglucosamine (GlcNAc, three-letter code NAG in the PDB). I was not able to use the `l-/d-` or `p/f` configurations. 

For example:
```
iupac_specific = "l-Galp(b1-4)GlcNAc"
print(gl.glycan(iupac_specific)) # works

modified_iupac = "Gal(b1-4)l-GlcNAc" # attempt to specify the specific stereoisomer
print(gl.glycan(modified_iupac)) # returns an error

modified_iupac_2 = "Gal(b1-4)NGZ" # with the PDB ID as before
print(gl.glycan(modified_iupac_2)) # works fine
```

Reading your new documentation, I see it mentions that specifying modified monosaccharides might be more challenging. In this case, is there a way to specify the `l-/d-` and `p/f` configurations while still using GlcNAc as the keyword?

**Comment by NoahHenrikKleinschmidt** [comment]
Sure thing, I'm happy you are using Glycosylator and finding some good things to improve or mention in the documentation 😄 

Yes, the whole `l-` vs `d-` notation can be a little tricky. The truth of the matter is that the PDB database is not 100% uniform when it comes to the identifiers that are stored, so it is possible that certain components simply have a (slightly) different identifier that you can/have to use to get them. Let's break it down for your case with the GlcNAc 😅 

> ### TLDR
> There are actually two formats that are supported by the PDB database a long and short one. Glycosylator is only designed to work with the long format but the short format can be used with some tweaking (not recommended though). However, in some cases, it may be quicker to identify the component you want to use in the database and directly specify the ID to avoid running into issues.

### Using long format: `{a|b}?-{l|d}-<code>{p|f}<mods>`
In the case of `NGZ` the identifier is specifically `a-l-GlcpNAc`. So, breaking it down, the system requires a leading `a-` identifier. Normally this information would be taken from the linkage specification that precedes the residue identifier but since the GlcNAc in this case is the first residue this information is missing (and would actually be ignored by the system anyway since most other sugars don't actually specify `a-<code>` as identifier but only `<code>`). Then the `p` comes directly after the monosaccharide code but before the specifications for modifications - this is actually a general pattern that I forgot to mention in the docs (I'll update that). So the pattern becomes `<code>{p|f}{mods}`, sounds fun right? In summary you can try to get the right component monosaccharides by constructing a pattern like so:

1. add `a-` or `b-` if it's the first residue, otherwise ignore since the linkage information is normally used for this. However, if the code does not find the component, try adding it anyway.
2. add `l-` or `d-` to specify the isomer
3. add the monosaccharide code, e.g. `Glc`, `Gal`, etc.
4. add `p` or `f` for the pyranose or furanose forms (in some cases only one may be available anyway but still specify)
5. add the code for modifiers, e.g. `NAc`

So, when I was providing the following iupac string the glycan did not through any errors:

```python
>>> iupac = "Gal(b1-4)a-l-GlcpNAc"
>>> print(gl.glycan(iupac))
Molecule None:
╙── Residue(GAL, 2)
    └── Residue(NGZ, 1)
```

A good practice, if you are unsure how to specify a specific compound in iupac format, is to use `has_compound` to see if Glycosylator is able to find a reference component for your specific input like so:

```python
>>> gl.has_compound("GlcNAc")
True
>>> gl.has_compound("l-GlcNAcp")
False
>>> gl.has_compound("l-GlcpNAc")
False
>>> gl.has_compound("a-l-GlcpNAc")
True
```

### Using short format: `{l|d}<code>{p|f}<mods>{a|b}`

<details>
<summary>Expand to read</summary>
There is actually a slightly "shorter" version of these identifiers also in the PDB database. Here, the pattern is slightly different in that the `a` or `b` is at the end of the string, and there are no `-` delimiting different sub-strings. The rest is more or less similar. However, the iupac string parsing is not designed to work with these short-hand formats since `a` or `b` data that is extracted from the linkage specifications is prepended to conform to the long format, so this kind of format will only work with (1) the first residue or (2) if you specify all linkages to be in `(a{}-{})` format and only specify the isomer as part of the residue specification. To make a little example: 

```python
>>> iupac = "dGalpNAc(b1-4)GlcNAc"
>>> gl.glycan(iupac)
# raises an error since no component for `b-dGalpNAc` is available

# but if we claim it's an alpha 1-4 linkage and still specify the beta isomer
# like so the code works...
>>> iupac = "dGalpNAcb(a1-4)GlcNAc"
>>> glc.glycan(iupac)
 Molecule None:
╙── Residue(NGA, 2)
    └── Residue(NDG, 1)
```

As you can see though this now starts to make the iupac strings very unconventional, so it's best to stick to the long format or even directly provide the component IDs if you know them. 
</details>

### Querying the database to find the right ID

<details>
<summary>Expand to read</summary>
The good thing is that you can query the database a little more freely so it is probably easier to figure out which component is the right one instead of figuring out which ID you should specify. I'll add some code to help query the database, just for completeness sake:

```python
# query the database using `get_compound`
# which returns a list of potential hits with data in dictionary form
>>> possible_hits = gl.get_compound("GlcNAc", return_type="dict")

# get rid of the pdb data we are not interested in to keep only the metadata
>>> possible_hits = [i[0] for i in possible_hits]

# now we can start investigating identifiers such as the registered names
>>> for hit in possible_hits:
            print(hit["id"])
            print("-" * 20)
            print("\n".join(hit["names"]))
            print("-" * 20)
```
</details>

So that wraps it up about the isomer-specific identifiers in iupac format. You can specify a isomers quite precisely but the PDB database is unfortunately not 100% uniform in the identifiers. Maybe we can get a bachelor or summer project student to make the identifiers more uniform manually at some point 😅 .  Hope that helps!
